cpp/src/hash/hashing.cu

```diff
@@ -77,84 +78,276 @@ class bitwise_partitioner
+            partitioner_type(num_partitions),
+            row_partition_numbers.data().get(),
+            block_partition_sizes.data().get(),
+            global_partition_sizes.data().get());
   }
 
-  // Scatter input rows to output partitions given the partition map
-  auto const partition_map = column_view(data_type(INT32), num_rows,
-    row_partition_numbers.data().get());
-  auto output = experimental::detail::scatter_to_tables(input, partition_map, mr, stream);
-
-  // Pad with empty tables if we have less than num_partitions
-  if (output.size() < static_cast<size_t>(num_partitions)) {
-    std::generate_n(std::back_inserter(output), num_partitions - output.size(),
-      [&input]() { return experimental::empty_like(input); });
-  }
-
-  return output;
+  // Compute exclusive scan of all blocks' partition sizes in-place to determine 
+  // the starting point for each blocks portion of each partition in the output
+  cudf::size_type * scanned_block_partition_sizes{block_partition_sizes.data().get()};
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         block_partition_sizes.begin(), 
+                         block_partition_sizes.end(), 
+                         scanned_block_partition_sizes);
+
+  // Compute exclusive scan of size of each partition to determine offset location
+  // of each partition in final output.
+  // TODO This can be done independently on a separate stream
+  size_type * scanned_global_partition_sizes{global_partition_sizes.data().get()};
+  thrust::exclusive_scan(rmm::exec_policy(stream)->on(stream),
+                         global_partition_sizes.begin(), 
+                         global_partition_sizes.end(),
+                         scanned_global_partition_sizes);
+
+  // Copy the result of the exlusive scan to the output offsets array
+  // to indicate the starting point for each partition in the output
+  std::vector<size_type> partition_offsets(num_partitions);
+  CUDA_TRY(cudaMemcpyAsync(partition_offsets.data(), 
+                           scanned_global_partition_sizes, 
+                           num_partitions * sizeof(size_type),
+                           cudaMemcpyDeviceToHost,
+                           stream));
+
+  // Compute the output location for each row in-place based on it's 
+  // partition number such that each partition will be contiguous in memory
+  size_type * row_output_locations{row_partition_numbers.data().get()};
+  compute_row_output_locations
+      <<<grid_size, BLOCK_SIZE, num_partitions * sizeof(size_type), stream>>>
+          (row_output_locations, num_rows, num_partitions, scanned_block_partition_sizes);
+
+  auto scatter_map = column_view{data_type{INT32}, num_rows, row_output_locations};
+  auto output = experimental::detail::scatter(input, scatter_map, input, false, mr, stream);
+
+  return std::make_pair(std::move(output), std::move(partition_offsets));
 }
 
+struct nvtx_raii {
```

Comment:
maybe add a comment to help understand what this is.

Reply:
Done